---
title: "Practical 02 SG: Hardy-Weinberg equilibrium"
author: "Write here your names and surnames"
date: "Hand-in: 30/11/2020"
output:
  pdf_document: 
    keep_tex: yes
    fig_crop: no
  html_document: default
---

Resolve the following exercise in groups of two students. Perform the computations and make the graphics that are asked for in the practical below. Take care to give each graph a title, and clearly label $x$ and $y$ axes, and to answer all questions asked. You can write your solution in a word or Latex document and generate a pdf file with your solution. Alternatively, you may generate a solution pdf file with Markdown. You can use R packages **data.table** and **HardyWeinberg** for the computations. Take care to number your answer exactly as in this exercise. Upload your solution in **pdf format** to the web page of the course at raco.fib.upc.edu  no later than the hand-in date.

```{r dependencies, echo=FALSE}
library(genetics)
library(HardyWeinberg)
library(data.table)
```

1. The file [TSIChr22v4.raw](http://www-eio.upc.es/~jan/data/bsg/TSIChr22v4.raw) contains genotype information of individuals from Tuscany in Italy, taken from the 1,000 Genomes project. The datafile contains all single nucleotide polymorphisms on chromosome 22 for which complete information is available. Load this data into the R environment. Use the `fread` instruction of the package `data.table`, which is more efficient for reading large datafiles. This data is in (0,1,2) format, where 0 and 2 represent the homozygotes AA and BB, and 1 represents the heterozygote AB. The first six leading columns of the data matrix can be ignored, as they do not contain any genetic information.

```{r 1st}

# Load data
dataset <- fread('TSIChr22v4.raw', data.table=FALSE)

# Remove leading columns and display data
dataset <- dataset[-c(1:6)]
head(dataset[1:8])

```


2. (1p) How many individuals does the database contain, and how many variants? What percentage of the variants is monomorphic? Remove all monomorphic SNPs from the database. How many variants remain in the database?  

```{r 2nd}

# Number of individuals and variants
n <- nrow(dataset)
v <- ncol(dataset)
cat(paste('Number of individuals:', n, '\n'))
cat(paste('Number of variants:', v, '\n'))

# Number of polymorphic variants
polymorphic_cols <- apply(dataset, 2, function(x) length(unique(x)) > 1)
dataset_poly <- dataset[polymorphic_cols]
vp <- ncol(dataset_poly)
cat(paste('Number of polymorphic variants:', vp, '\n'))

```

3. (3p) Extract polymorphism rs587756191_T from the datamatrix, and determine its genotype counts. Apply a chi-square test for Hardy-Weinberg equilibrium, with and without continuity correction. Also try an exact test, and a permutation test. You can use function `HWChisq`, 
`HWExact` and `HWPerm` for this purpose. Do you think this variant is in equilibrium? Argue your answer. 

```{r 3rd, echo=FALSE} 

# Genotype counts function
count_genotypes <- function(variant) {
  counts <- c(sum(variant==0), sum(variant==1), sum(variant==2))
  names(counts) = c("AA", "AB", "BB")
  counts
}

# Counting genotypes of 'rs587756191_T'
variant <- dataset_poly[,c('rs587756191_T')]
counts <- count_genotypes(variant)

# HW equilibrium tests
HWChisq(counts)
HWChisq(counts, cc=0)
HWExact(counts)
HWPerm(counts)
```

If the p-value is small enough, we reject the null hypotesis that the data comes from the specified distribution. We have to keep in mind that small number of counts of one genotype may lead to incorrect test results. In our case, we have **106** AA genotypes, **1** AB genotype and no BB genotypes. Chi-square tests warn us that the results may be incorrect because of the mentioned reason. Exact test and permutation test give better results with p-value equal to 1. So, considering these value, we can say that this variant is in equilibrium.

4. Determine the genotype counts for all these variants, and store them in a $p \times 3$ matrix.

```{r 4th}

genotype_counts <- data.frame(t(apply(dataset_poly, 2, count_genotypes)))
genotype_counts[1:20,]

```

5. (1p) Apply a chi-square test without continuity correction for Hardy-Weinberg equilibrium to each SNP. You can use `HWChisqStats` for this purpose. How many SNPs are significant (use $\alpha = 0.05$)? 

```{r 5th}

pvalues_chi <- HWChisqStats(genotype_counts, pvalues=TRUE)
pvalues_significant_chi <- pvalues_chi[pvalues_chi <= 0.05]
cat(paste('There are', length(pvalues_significant_chi), 'significant SNPs out of', length(pvalues_chi)))

```

6.  (1p) How many markers of the remaining non-monomorphic markers would you expect to be out of equilibrium by the effect of chance alone? 

```{r 6th}

chance = length(pvalues_significant_chi) / length(pvalues_chi )
chance

cat(paste('We would expect to see', round(chance * 100, digits = 3), '% of markers to be out of equilibrium. '))
```

7.  (2p) Which SNP is most significant according to the chi-square test results? Give it genotype counts. In which sense is this genotypic composition unusual?

```{r 7th}

most_significant_index <- which.min(pvalues_chi)
genotype_counts[most_significant_index,]

```
It is unusual because all genotypes of this SNPs are equal to AA besides one which is BB.

8. (1p) Apply an Exact test for Hardy-Weinberg equilibrium to each SNP. You can use function `HWExactStats` for fast computation. How many SNPs are significant (use $\alpha = 0.05$). Is the result consistent with the chi-square test? 

```{r 8th}

pvalues_exact <- HWExactStats(genotype_counts)
pvalues_significant_exact <- pvalues_exact[pvalues_exact <= 0.05]
cat(paste('There are', length(pvalues_significant_exact), 'significant SNPs out of', length(pvalues_exact)))

```
The number of significant SNPs given by an Exact test is smaller than the one given by a chi-square test, but if we compare it to the total number of SNPs, we can say that it is similar.

9. (2p) Which SNP is most significant according to the exact test results? Give its genotype counts. In which sense is this genotypic composition unusual?

```{r 9th}

most_significant_index <- which.min(pvalues_exact)
genotype_counts[most_significant_index,]

```
This genotypic composition is unusual because it has almost equal number of AA and BB genotypes, but none of AB genotypes.

10. (1p) Apply a likelihood ratio test for Hardy-Weinberg equilibrium to each SNP, using the `HWLratio` function. How many SNPs are significant
(use $\alpha = 0.05$). Is the result consistent with the chi-square test? 

```{r 10th}

pvalues_likelihood <- apply(genotype_counts, 1, function(x) HWLratio(x, verbose=FALSE)$pval)
pvalues_significant_likelihood <- pvalues_likelihood[pvalues_likelihood <= 0.05]
cat(paste('There are', length(pvalues_significant_likelihood), 'significant SNPs out of', length(pvalues_likelihood)))

```

The result of the likelihood ratio test is much closer to chi-square test then exact test was. Likelihood ratio test estimated 7955 significant values, and chi-square test estimated 8162 of them.

11. (1p) Apply a permutation test for Hardy-Weinberg equilibrium to the first 10 SNPs, using the classical chi-square test (without continuity correction) as a test statistic. 
List the 10 p-values, together with the 10 p-values of the exact tests. Are the result consistent? 

```{r 11th}

pvalues_perm_head <- apply(genotype_counts[1:10,], 1, function(x) HWPerm(x, verbose=FALSE)$pval)
pvalues_exact_head <- pvalues_exact[1:10]
rbind(pvalues_perm_head, pvalues_exact_head)

```

Not all p-values are consisent. We can see that all besides p-values of variant **rs62224609_C** are close.

12. (1p) Depict all SNPs simultaeneously in a ternary plot with function `HWTernaryPlot` and comment on your result (because many genotype counts repeat, you may use `UniqueGenotypeCounts` to speed up the computations)

```{r 12th}

unique_genotype_counts <- UniqueGenotypeCounts(genotype_counts)[1:3]
HWTernaryPlot(unique_genotype_counts)

```

13. (1p) Can you explain why half of the ternary diagram is empty? 

Discussing the possibilities why the right part of the ternary diagram is empty we came to the conclusion that A is probably always taken as the more frequent allele and thus there is no way when BB could be a more frequent case than AA.

14. (2p) Make a histogram of the $p$-values obtained in the chi-square test. What distribution would you expect if HWE would hold for the data set? Make a Q-Q plot of the $p$ values obtained in the chi-square test against the quantiles of the distribution that you consider relevant. What is your conclusion?. 

```{r 14th}

hist(pvalues_chi, freq=FALSE)
curve (dbeta(x, shape1 = 80, shape2 = 3), add=TRUE, col="red")

x_qbeta <- seq(0, 1, length.out = length(pvalues_chi))
y_qbeta <- rbeta(x_qbeta, shape1 = 80, shape2 = 3)

qqplot(pvalues_chi, y_qbeta, main="Q-Q plot", xlab = "P-values", ylab = "Beta distribution quantiles")

```

If HWE would hold for the whole data set, we would expect that the left part of the histogram (below 0.05) is 0, i.e. that no row has p-value below 0.05. Ideally we would expect the mass of the histogram to shift as far right as possible.

This histogram is shaped like some very skewed distribution with heavy left tail. We could not find some distribution with a heavy tail like this, do the best fit for this is beta distribution with parameters alpha = 80 and beta = 3. Q-Q plot is presented in the plot, we see that in the middle there is linear behavior, but coming closer to each end of the curve the points start to deviate significantly from the line.

15. (1p) Imagine that for a particular marker the counts of the two homozygotes are accidentally interchanged. Would this affect the statistical tests for HWE? Try it on the computer if you want. Argue your answer.

```{r 15th}
counts <- c(AA=40, AB=10, BB=15)
HWChisq(counts)

interchanged_counts <- c(AA=15, AB=10, BB=40)
HWChisq(interchanged_counts)
```

If counts of two homozygotes were accidentally interchanged, it wouldn't affect the statistical test for HWE. Switching counts of two homozygotes also switches the probabilities p and q, which does not change the test value. It can be seen if we observe formula for Chi-squared test (on slide 19 from the lesson's PDF). Heterozygote count does not change, so the nominator stays equal. Denominator consists of factor *p-squared*, *q-squared* and *n* and stays equal if we switch p and q values.


16. (3p) Compute the inbreeding coefficient ($\hat{f}$) for each SNP, and make a histogram of $\hat{f}$. You can use function `HWf` for this purpose. Give descriptive statistics (mean, standard deviation, etc) of $\hat{f}$ calculated over the set of SNPs. What distribution do you expect $\hat{f}$ to follow theoretically? Use a probability plot to confirm your idea.

```{r 16th}
fs <- HWf(as.matrix(genotype_counts))
hist(fs)
cat(paste('Mean:', mean(fs), '\nStandard deviation:', sd(fs), '\nMedian:', median(fs), '\nMin:', min(fs), '\nMax:', max(fs)))
```

17. (2p) Make a plot of the observed chi-square statistics against the inbreeding coefficient ($\hat{f}$). What do you observe? Can you give an equation that relates the two statistics? 

```{r 17th}

plot(pvalues_chi, fs)

```
We can observe that p-values are closer to zero as inbreeding coefficient is further from zero, and vice versa. That makes sense, because low p-value indicates that variant is not in equilibrium, and it happens when there are no heterozygotes (f = 1) or much more heterozygotes than homozygotes (f = -1).

18. (2p) We reconsider the exact test for HWE, using different significant levels. Report the number and percentage of significant variants using an exac test for HWE with $\alpha = 0.10, 0.05, 0.01$ and 0.001. State your conclusions.

```{r 18th}

get_significant_percentage <- function(pvalues, alpha) {
  significant <- pvalues[pvalues <= alpha]
  length(significant) / length(pvalues)
}

significant_percentages <- c(get_significant_percentage(pvalues_exact, 0.10),
                             get_significant_percentage(pvalues_exact, 0.05),
                             get_significant_percentage(pvalues_exact, 0.01),
                             get_significant_percentage(pvalues_exact, 0.001))
names(significant_percentages) <- c('0.1', '0.05', '0.01', '0.001')
significant_percentages

```

We can see that percentage of significant variants decreases as alpha is approaching zero.
