---
title: "Practical 03 SG: Haplotype estimation"
author: "Lovro KataliniÄ‡ and Ivan Almer"
date: "Hand-in: 05/12/2020"
output:
  pdf_document: 
    keep_tex: yes
    fig_crop: no
  html_document: default
---

Resolve the following exercise in groups of two students. Perform the computations and make the graphics that are asked for in the practical below. Take care to give each graph a title, and clearly label $x$ and $y$ axes, and to answer all questions asked. You can write your solution in a word or Latex document and generate a pdf file with your solution. Alternatively, you may generate a solution pdf file with Markdown. You can use R packages **genetics**, **haplo.stats**, **LDheatmap** and others for the computations. 
Take care to number your answer exactly as in this exercise. Upload your solution in **pdf format** to the web page of the course at raco.fib.upc.edu  no later than the hand-in date.

```{r dependencies, echo=FALSE}
library('genetics')
library('haplo.stats')
library('LDheatmap')
library('data.table')
```

1. Apolipoprotein E (APOE) is a protein involved in Alzheimer's disease. The corresponding gene _APOE_ has been mapped to chromosome 19. The file [APOE.dat](http://www-eio.upc.es/~jan/data/bsg/APOE.dat) contains genotype information of unrelated individuals for a set of SNPs in this gene. Load this data into the R environment. [APOE.zip](http://www-eio.upc.es/~jan/data/bsg/APOE.zip) contains the corresponding `.bim`, `.fam` and `.bed` files. You can use the `.bim` file to obtain information about the alleles of each polymorphism. 

```{r 1st}
apoe <- fread('APOE.dat', data.table=FALSE)
rownames(apoe) <- apoe[,1]
apoe <- apoe[,-c(1)]
head(apoe)
```

2. (1p) How many individuals and how many SNPs are there in the database? What percentage of the data is missing? 

```{r 2nd}
n <- nrow(apoe)
p <- ncol(apoe)
cat(paste('There are', n, 'individuals and', p, 'SNPs in the database.\n'))

missing_percentage <- 100*sum(as.integer(is.na(apoe)))/(n*p)
cat(paste(missing_percentage, '% of data is missing.'))
```

3. (1p) Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set? 

```{r 3rd}
possible_haplotypes <- 2^p
cat(paste('For bi-allelic SNPS, theoretically 2^', p, ' = ', format(possible_haplotypes, digits=3), ' haplotypes can be found for this data set.', sep=''))
```

4. (2p) Estimate haplotype frequencies using the haplo.stats package (set the minimum posterior probability to 0.001). How many haplotypes do you find? List the estimated probabilities in decreasing order. Which haplotype number is the most common?

```{r em-function-preparation} 
get_prepared_data <- function(data) {
  prepared_data <- c()
  for(i in 1:ncol(data)) {
    prepared_data <- cbind(prepared_data, 
                           substr(data[,i],1,1),
                           substr(data[,i],3,3))
  }
  prepared_data
}
```

```{r 4th}
estimate_haplotypes <- function(data) {
  data_prepared <- get_prepared_data(data)
  haplo_estimation <- haplo.em(data_prepared,
                               locus.label=colnames(data),
                               control=haplo.em.control(min.posterior=1e-3))
  
  haplos <- haplo_estimation$haplotype
  haplo_num <- dim(haplos)[1]
  cat(paste('Algorithm found', haplo_num, 'haplotypes.\n\n'))
  
  haplo_probs <- haplo_estimation$hap.prob
  haplo_probs_ordered <- order(haplo_probs, decreasing=TRUE)
  cat('Estimated probabilities in decreasing order:\n')
  cat(haplo_probs[haplo_probs_ordered])
  cat(paste('\n\nMost common is haplotype numbered ', haplo_probs_ordered[1], '.', sep=''))
  
  return(haplo_estimation)
}

apoe_haplos <- estimate_haplotypes(apoe)
```

5.  (2p) Is the haplotypic constitution of any of the individuals in the database ambiguous or uncertain?  For how many? What is the most likely haplotypic constitution of individual NA20763? (identify the constitution by the corresponding haplotype numbers).

```{r 5th}
constitutions <- apoe_haplos$nreps
ambiguous <- sum(constitutions > 1)
cat(paste('There are ', ambiguous, ' ambigous haplotypic constituions of an individual.', sep=''))

individual_index <- which(rownames(apoe) == 'NA20763')
most_likely_hc <- apoe_haplos$hap1code[individual_index]
cat(paste('\nMost likely haplotypic constitution of individual NA20763 is ', most_likely_hc, '.', sep=''))
```

6. (1p) Suppose we would delete polymorphism rs374311741 from the database prior to haplotype estimation. Would this affect the results obtained? Justify your answer.

```{r 6th}
apoe_without_polymorphism <- subset(apoe, select=-c(rs374311741))
apoe_without_polymorphism_haplo <- estimate_haplotypes(apoe_without_polymorphism)
```

Deleting one column from the database resulted in minor changes in haplotype probabilities, but the haplotype count stays the same. (justify)

7. (1p) Remove all genetic variants that have a minor allele frequency below 0.10 from the database, and re-run `haplo.em`. How does this affect the number of haplotypes?

```{r 7th}
maf <- function(x){
  x <- genotype(x,sep="/")
  out <- summary(x)
  af1 <- min(out$allele.freq[,2],na.rm=TRUE)
  af1[af1==1] <- 0
  af1
}

mafs <- apply(apoe, 2, maf)
apoe_filtered <- apoe[, mafs > 0.10]
cat(paste('By filtering genetic variants that have MAF below 0.1, we reduced their number from ', p, ' to ', dim(apoe_filtered)[2], '.\n\n', sep=''))

apoe_filtered_haplo <- estimate_haplotypes(apoe_filtered)
```

The results changed dramatically when variants with minor alelle frequency below 0.10 were removed. With filtered dataset like this one, function `haplo.em` found 8 haplotypes.

8. (2p) We could consider the newly created haplotypes in our last run of `haplo.em` as the alleles of a new superlocus. Which is, under the assumption of Hardy-Weinberg equilibrium, the most likely genotype at this new locus? What is the probability of this genotype? Which genotype is the second most likely, and what is its probability?

```{r 8th}
haplotypes <- apoe_filtered_haplo$haplotype
superlocus <- c()
for (i in 1:dim(haplotypes)[2]) {
  superlocus <- paste(superlocus, haplotypes[,i], sep='')
}
superlocus
summary(genotype(superlocus, sep=''))
```

